src/matrixrtc/RoomAndToDeviceKeyTransport.ts

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2025 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import type { MatrixClient } from "../client.ts";
+import { logger as rootLogger, type Logger } from "../logger.ts";
+import { KeyTransportEvents, type KeyTransportEventsHandlerMap, type IKeyTransport } from "./IKeyTransport.ts";
+import { type CallMembership } from "./CallMembership.ts";
+import { RoomKeyTransport } from "./RoomKeyTransport.ts";
+import type { Statistics } from "./types.ts";
+import { ToDeviceKeyTransport } from "./ToDeviceKeyTransport.ts";
+import { TypedEventEmitter } from "../models/typed-event-emitter.ts";
+import { type Room } from "../models/room.ts";
+
+export interface EnabledTransports {
+    toDevice: boolean;
+    room: boolean;
+}
+
+export enum RoomAndToDeviceEvents {
+    EnabledTransportsChanged = "enabled_transports_changed",
+}
+export type RoomAndToDeviceEventsHandlerMap = {
+    [RoomAndToDeviceEvents.EnabledTransportsChanged]: (enabledTransports: EnabledTransports) => void;
```

Comment:
As said I think this could be avoidded by caching the current key and replay on transport change?

Reply:
The reason for the emission is not only the resending but mostly that we can update the label in ec

Reply:
Also this would make us loose the retry mechanisms from the Encryption manager.
SendKey for the toDevice would succeed, we cache the key, we need to switch to a new transport and we send the key over the new transport. if that key sending fails there is nothing we can do. Except we build another retrying mechanism into the transport.

Reply:
discussed sync: We keep the event emissino while we use `RoomAndToDeviceTransport` as a stop gap.